src/server/transaction.cc

```diff
@@ -702,30 +694,49 @@ void Transaction::ScheduleInternal() {
   DCHECK_GT(unique_shard_cnt_, 0u);
   DCHECK(!IsAtomicMulti() || cid_->IsMultiTransactional());
 
-  DVLOG(1) << "ScheduleInternal " << cid_->name() << " on " << unique_shard_cnt_ << " shards";
+  // Try running immediately if we're concluding and either:
+  // - have a single shard, and thus never have to cancel due to reordering
+  // - run as an idempotent command, meaning we can safely repeat the operation
+  bool can_run_immediate = !IsGlobal() && (coordinator_state_ & COORD_CONCLUDING) &&
+                           (unique_shard_cnt_ == 1 || (cid_->opt_mask() & CO::IDEMPOTENT));
+
+  DVLOG(1) << "ScheduleInternal " << cid_->name() << " on " << unique_shard_cnt_ << " shards "
+           << " immediate run: " << can_run_immediate;
 
   auto is_active = [this](uint32_t i) { return IsActive(i); };
 
   // Loop until successfully scheduled in all shards.
   while (true) {
     stats_.schedule_attempts++;
 
-    txid_ = op_seq.fetch_add(1, memory_order_relaxed);
+    // This is a contention point for all threads - avoid using it unless necessary.
+    // Single shard operations can assign txid later if the immediate run failed.
```

Comment:
we need to add somewhere in the code or maybe the tx doc what "immediate" run  is.

Reply:
A few lines above, expanded the comment